panther_driver/src/driver_node.py

```diff
@@ -51,273 +32,320 @@ def euler_to_quaternion(yaw, pitch, roll):
-    rospy.loginfo(f"[{rospy.get_name()}] Start with creating a network representing one CAN bus")
-    network = canopen.Network()
-    rospy.loginfo(f"[{rospy.get_name()}] Add some nodes with corresponding Object Dictionaries")
-    front_controller = canopen.RemoteNode(1, eds_file)
-    rear_controller = canopen.RemoteNode(2, eds_file)
-    network.add_node(front_controller)
-    network.add_node(rear_controller)
-    rospy.loginfo(f"[{rospy.get_name()}] Connect to the CAN bus")
-    network.connect(channel=can_interface, bustype="socketcan")
-
-    # rospy.loginfo(f"[{rospy.get_name()}] Reset encoders")
-    # front_controller.sdo["Cmd_SENCNTR"][1].raw = 0
-    # front_controller.sdo["Cmd_SENCNTR"][2].raw = 0
-
-    robot_x_pos = 0.0
-    robot_y_pos = 0.0
-    robot_th_pos = 0.0
-
-    wheel_pos = [0.0, 0.0, 0.0, 0.0]
-    wheel_vel = [0.0, 0.0, 0.0, 0.0]
-    wheel_curr = [0.0, 0.0, 0.0, 0.0]
-
-    last_time = rospy.Time.now()
-
-    # VARIABLE FOR ERROR TRACKING
-    err_count = 0
-
-    while not rospy.is_shutdown():
+class PantherDriver(object):
+    def __init__(self, name) -> None:
+        rospy.init_node(name, anonymous=False)
+
+        assert rospy.has_param(
+            "~eds_file"
+        ), f"[{rospy.get_name()}] eds_file not defined, can not start CAN interface"
+        self.eds_file = rospy.get_param("~eds_file")
+        self.can_interface = rospy.get_param("~can_interface", "panther_can")
+        self.kinematics_type = rospy.get_param("~wheel_type", "differential")
+        self.motor_torque_constant = rospy.get_param("~motor_torque_constant", 2.6149)
+
+        self.publish_tf = rospy.get_param("~publish_tf", True)
+        self.publish_odom = rospy.get_param("~publish_odometry", True)
+        self.publish_pose = rospy.get_param("~publish_pose", True)
+        self.publish_joints = rospy.get_param("~publish_joints", True)
+
+        self.odom_frame = rospy.get_param("~odom_frame", "odom")
+        self.base_link_frame = rospy.get_param("~base_link_frame", "base_link")
+        self.wheels_joints_names = [
+            "front_left_wheel_joint",
+            "front_right_wheel_joint",
+            "rear_left_wheel_joint",
+            "rear_right_wheel_joint",
+        ]
+
+        self.pose_msg = Pose()
+        self.tf_broadcaster = tf2_ros.TransformBroadcaster()
+        self.tf_stamped = TransformStamped()
+        self.tf_stamped.header.frame_id = self.odom_frame
+        self.tf_stamped.child_frame_id = self.base_link_frame
+        self.joint_state_msg = JointState()
+        self.joint_state_msg.header.frame_id = self.base_link_frame
+        self.joint_state_msg.name = self.wheels_joints_names
+        self.odom_msg = Odometry()
+        self.odom_msg.header.frame_id = self.odom_frame
+        self.driver_state_msg = DriverStateArr()
+
+        self.cmd_vel_timeout = 0.2
+        self.main_timer_freq = 1./15.           # 15 Hz
+        self.driver_state_timer_freq = 1./4.    # 4 Hz
```

Comment:
```suggestion
        self.main_timer_period = 1.0 / 15.0           # 15 Hz
        self.driver_state_timer_period = 1.0 / 4.0    # 4 Hz
```
Names suggest its frequency, while it's timer period + formatting. Also, it's worth considering adding rosparms for frequency values

Reply:
Good point, but in my opinion rosparams are dangerous because it may cause too much load on the CAN bus or too little time to execute the callback.